### PR TITLE
fix: update limit for es response

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-3986](https://2u-internal.atlassian.net/browse/PROD-3986)
Adds a setting to `update_index`. This will increase the `max_result_window` for all existing indexes and subsequently update any newly created index as well. 

~Also updates the unit test that previously didn't support a custom value of `max_result_window` in ES related logic.~
Keeping the same limit count for unit tests as we won't reach this limit while testing.

### CODE REVIEW

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py [hunk 1]
```diff
@@ -127,6 +127,8 @@ def _update(self, models, options):
         for index_alias_mapper in alias_mappings:
             index_alias_mapper.registered_index._name = index_alias_mapper.alias  # pylint: disable=protected-access
 
+        conn.indices.put_settings(body={"index": {"max_result_window": settings.MAX_RESULT_WINDOW}})
+
```
Comment (DawoudSheraz) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py]: If this is global setting, it should be set only once. it should not be set in a periodically running management command.

Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py]: This setting is added to each index separately. Having it in this place specifically works as all the new/updated indexes will have this setting.

**Review by zawan-ila — COMMENTED** [review]
Why do we need it in 2 management commands? Wouldn't the settings update when only one runs? Would it be nicer to extract this into a separate management command?

**Comment by Ali-D-Akbar** [comment]
> Why do we need it in 2 management commands? 

To update this setting for both cases where we're either updating (through update_index) or creating new/rebuilding indexes (through search_index).

> Wouldn't the settings update when only one runs? 

In case when we create new index, it won't have the settings. These settings work for pre-existing indexes.

>Would it be nicer to extract this into a separate management command?

That's debatable although the same command will be needed to run in both of the commands above.

File: course_discovery/apps/core/tests/mixins.py [hunk 2]
```diff
@@ -17,6 +18,10 @@ class ElasticsearchTestMixin:
     def setUp(self):
         super().setUp()
         self.es = get_connection()
+        try:
+            self.es.indices.put_settings(body={"index": {"max_result_window": settings.MAX_RESULT_WINDOW}})
+        except NotFoundError:
+            pass
```
Comment (DawoudSheraz) [inline course_discovery/apps/core/tests/mixins.py]: I personally am not a fan of this, specially in unit tests. Its a recipe for weirdness and unexpected behavior.

File: course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py [hunk 3]
```diff
@@ -8,6 +9,7 @@
 from course_discovery.apps.course_metadata.tests.factories import CourseFactory, CourseRunFactory
 
 
+@pytest.mark.django_db
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py]: why is this needed? I am assuming APITestCase is child of django or DRF's TestCase which takes care of Django DB ops.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py [hunk 4]
```diff
@@ -60,6 +60,11 @@ def _create(self, models, aliases, options):
                     )
                 )
                 ElasticsearchUtils.set_alias(es_connection, created_index_info.alias, created_index_info.name)
+        try:
+            es_connection = get_connection()
+            es_connection.indices.put_settings(body={"index": {"max_result_window": settings.MAX_RESULT_WINDOW}})
+        except NotFoundError:
+            pass
```
Comment (DawoudSheraz) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: In what cases will NotFoundError raise?

Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: Just in case if no index is created, we won't be adding this setting to any index.

File: course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py]: True, it can be removed. 👍

File: course_discovery/apps/core/tests/mixins.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/core/tests/mixins.py]: Basically, no setting is added when there is no index.

File: course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py (hunk 3 above)
Comment (DawoudSheraz) [reply course_discovery/apps/api/v1/tests/test_views/test_catalog_queries.py]: Well, go ahead then.

File: course_discovery/apps/core/tests/mixins.py (hunk 2 above)
Comment (DawoudSheraz) [reply course_discovery/apps/core/tests/mixins.py]: under what Testcases is this valid?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py (hunk 4 above)
Comment (DawoudSheraz) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: How so? This is being done inside _create. Under what circumstances will we not create an index despite calling create?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py (hunk 1 above)
Comment (DawoudSheraz) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py]: Is the operation idempotent?

File: course_discovery/apps/core/tests/mixins.py (hunk 2 above)
Comment (zawan-ila) [reply course_discovery/apps/core/tests/mixins.py]: I don't really see any tests for this. Not sure why this should be required in the first place?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py (hunk 4 above)
Comment (zawan-ila) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: It'd be nice to put this logic inside the `ElasticSearchUtils` class. Then you might be able to get by by calling it within `ElasticsearchUtils.create_index`.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py [hunk 5]
```diff
@@ -127,6 +127,8 @@ def _update(self, models, options):
         for index_alias_mapper in alias_mappings:
             index_alias_mapper.registered_index._name = index_alias_mapper.alias  # pylint: disable=protected-access
 
+        conn.indices.put_settings(body={"index": {"max_result_window": settings.MAX_RESULT_WINDOW}})
```
Comment (zawan-ila) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py]: Similar to my comment above, can we place this inside `ElasticsearchUtils`?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py (hunk 1 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py]: Yes, it is.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: @zawan-ila can you elaborate on where I should put these settings in the utils?

File: course_discovery/apps/api/tests/test_serializers.py [hunk 6]
```diff
@@ -2624,8 +2624,7 @@ def get_expected_data(cls, person, request):
 
 
 @pytest.mark.django_db
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/tests/test_serializers.py]: You might not need django_db mark if you are already using Django testcase.

File: course_discovery/apps/core/tests/mixins.py (hunk 2 above)
Comment (DawoudSheraz) [reply course_discovery/apps/core/tests/mixins.py]: For some testing, can you try this in setUpclass? I still dont understand why this is needed for tests as there are not any items in unit tests thaa can cross the threshold.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py (hunk 4 above)
Comment (DawoudSheraz) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: This is what @zawan-ila is referring to https://github.com/openedx/course-discovery/blob/master/course_discovery/apps/core/utils.py#L28. When calling create_index, this result_window update can be done there (possibly without try/catch?).

File: course_discovery/apps/core/tests/mixins.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/core/tests/mixins.py]: The settings needs to be updated bcz I changed `ELASTICSEARCH_DSL_QUERYSET_PAGINATION`, `ELASTICSEARCH_DSL_LOAD_PER_QUERY` to 15000 in `base.py`.
Should I set this to 10k for tests only?

Comment (DawoudSheraz) [reply course_discovery/apps/core/tests/mixins.py]: Hmmm, yeah I think we can keep the settings as it is for tests and not complicate this unnecessarily.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/search_index.py]: I tried adding this settings isn't being added for all indexes. Seems like it needs to be added after all the indexes and aliases are defined.

**Review by DawoudSheraz — APPROVED** [review]

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py [hunk 7]
```diff
@@ -117,11 +117,13 @@ def _update(self, models, options):
                     )
                     if record_count_is_sane:
                         ElasticsearchUtils.set_alias(conn, alias, new_index_name)
+                        ElasticsearchUtils.update_max_result_window(conn, settings.MAX_RESULT_WINDOW, new_index_name)
```
Comment (zawan-ila) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py:120]: Would be really nice if we could do this just once at the start of the loop.

**Review by zawan-ila — APPROVED** [review]

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py (hunk 7 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py:120]: Should be done only after `set_alias`.
